include/sys/dmu_objset.h

```diff
@@ -125,6 +126,12 @@ struct objset {
 	kmutex_t os_user_ptr_lock;
 	void *os_user_ptr;
 	sa_os_t *os_sa;
+
+	/* kernel thread to upgrade this dataset */
+	kthread_t *os_upgrade_thread;
+	kmutex_t os_upgrade_lock;
+	kcondvar_t os_upgrade_cv;
+	boolean_t os_upgrade_exit;
```

Comment:
For 1000 active dataset - is it imaginary or real case - if this is the real case we need to solve it absolutely.

Actually I was going to make a proposal to start upgrade thread in ```dmu_objset_own()``` and kill them in ```dmu_objset_disown()``` but if this is the case we need to solve in some way else.